File: group_anagrams.cpp
// Time Complexity : NKlogK
// Space Complexity : O(NK)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : N/A

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> map;
        for(auto& string : strs) // O(N)
        {
            // store the sting in local variable and sort the key
            std::string key = string; 
            std::sort(key.begin(), key.end()); // KlogK

            // push it into the value vector
            map[key].push_back(string);
        }

        vector<vector<string>> result;
        for(auto& entry : map)
        {
            result.push_back(entry.second);
        }

        return result;
    }
};

File: GroupAnagrams.cs
public class Solution {
    // Time Complexity: O(n * k log k), where n is the number of strings and k is the average length of each string.
    // Space Complexity: O(n * k), where n is the number of strings and k is the average length of each string.
    
    public IList<IList<string>> GroupAnagrams(string[] strs) {
        Dictionary<string, IList<string>> anagramGroups = new Dictionary<string, IList<string>>();

        foreach (string str in strs)
        {
            char[] charArray = str.ToCharArray();
            Array.Sort(charArray);
            string sortedStr = new string(charArray);

            if (!anagramGroups.ContainsKey(sortedStr))
            {
                anagramGroups[sortedStr] = new List<string>();
            }

            anagramGroups[sortedStr].Add(str);
        }

        return new List<IList<string>>(anagramGroups.Values);
    }
}